### PR TITLE
Stream p2p agg attestation

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
Slasher should be receiving all the attestations it can from the beacon node it is connected to therefore aggregated attestations from p2p network that are validated should be sent to it as well

**Which issues(s) does this PR fix?**

Part of #5031

### CODE REVIEW

File: beacon-chain/sync/subscriber_beacon_aggregate_proof.go [hunk 1]
```diff
@@ -5,6 +5,9 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/feed"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/feed/operation"
+
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_beacon_aggregate_proof.go:10]:
```suggestion
```

File: beacon-chain/sync/subscriber_beacon_aggregate_proof.go [hunk 2]
```diff
@@ -22,5 +25,13 @@ func (r *Service) beaconAggregateProofSubscriber(ctx context.Context, msg proto.
...
 
+	// Broadcast the aggregated attestation on a feed to notify other services in the beacon node
+	// of a received aggregated attestation.
+	r.attestationNotifier.OperationFeed().Send(&feed.Event{
+		Type: operation.AggregatedAttReceived,
+		Data: &operation.AggregatedAttReceivedData{
+			Attestation: a.Message,
+		},
+	})
 	return r.attPool.SaveAggregatedAttestation(a.Message.Aggregate)
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_beacon_aggregate_proof.go:36]:
```suggestion

	return r.attPool.SaveAggregatedAttestation(a.Message.Aggregate)
```

File: beacon-chain/sync/subscriber_beacon_aggregate_proof_test.go [hunk 3]
```diff
@@ -5,6 +5,8 @@ import (
 	"reflect"
 	"testing"
 
+	mock "github.com/prysmaticlabs/prysm/beacon-chain/blockchain/testing"
+
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_beacon_aggregate_proof_test.go:9]:
```suggestion
```

File: beacon-chain/rpc/beacon/attestations.go [hunk 4]
```diff
@@ -250,6 +250,22 @@ func (bs *Server) StreamIndexedAttestations(
 				}
 				bs.ReceivedAttestationsBuffer <- data.Attestation
 			}
+			if event.Type == operation.AggregatedAttReceived {
+				data, ok := event.Data.(*operation.AggregatedAttReceivedData)
+				if !ok {
+					// Got bad data over the stream.
+					continue
+				}
+				if data.Attestation == nil {
```
Comment (terencechain) [inline beacon-chain/rpc/beacon/attestations.go:259]: Optional:

```suggestion
		if data.Attestation == nil || data.Attestation.Aggregate == nil {
```

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go]: thanks, changed

File: beacon-chain/rpc/beacon/attestations.go [hunk 5]
```diff
@@ -250,6 +250,18 @@ func (bs *Server) StreamIndexedAttestations(
...
+				data, ok := event.Data.(*operation.AggregatedAttReceivedData)
+				if !ok {
+					// Got bad data over the stream.
+					continue
+				}
+				if data.Attestation == nil || data.Attestation.Aggregate == nil {
+					// One nil attestation shouldn't stop the stream.
+					continue
+				}
+				bs.CollectedAttestationsBuffer <- []*ethpb.Attestation{data.Attestation.Aggregate}
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:263]: Can you change this to `ReceivedAttestationsBuffer` please, if its `Collected` then the slasher will get spammed with every attestation received.

File: beacon-chain/rpc/beacon/attestations.go [hunk 6]
```diff
@@ -249,6 +249,17 @@ func (bs *Server) StreamIndexedAttestations(
 					continue
 				}
 				bs.ReceivedAttestationsBuffer <- data.Attestation
+			} else if event.Type == operation.AggregatedAttReceived {
+				data, ok := event.Data.(*operation.AggregatedAttReceivedData)
+				if !ok {
+					// Got bad data over the stream.
```
Comment (prestonvanloon) [inline beacon-chain/rpc/beacon/attestations.go:255]: Can we log an error or warning? Seems like the stream could be easily broken and failing silently.

File: beacon-chain/rpc/beacon/attestations.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go]: can we revert it back?

File: beacon-chain/rpc/beacon/attestations.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go:255]: sure. adding

File: beacon-chain/rpc/beacon/attestations.go [hunk 7]
```diff
@@ -237,20 +237,41 @@ func (bs *Server) StreamIndexedAttestations(
 	go bs.collectReceivedAttestations(stream.Context())
 	for {
 		select {
-		case event := <-attestationsChannel:
+		case event, ok := <-attestationsChannel:
+			if !ok {
+				logrus.Error("Stream indexed attestations attestations channel is got closed")
```
Comment (rauljordan) [inline beacon-chain/rpc/beacon/attestations.go:242]:
```suggestion
				logrus.Error("Stream indexed attestations channel closed")
```

File: beacon-chain/rpc/beacon/attestations.go [hunk 8]
```diff
@@ -237,20 +237,41 @@ func (bs *Server) StreamIndexedAttestations(
...
 			if event.Type == operation.UnaggregatedAttReceived {
 				data, ok := event.Data.(*operation.UnAggregatedAttReceivedData)
 				if !ok {
 					// Got bad data over the stream.
+					logrus.Warningf("Got data of wrong type on stream expected *UnAggregatedAttReceivedData, received %T", event.Data)
 					continue
 				}
 				if data.Attestation == nil {
 					// One nil attestation shouldn't stop the stream.
+					logrus.Info("Got nil attestation")
```
Comment (rauljordan) [inline beacon-chain/rpc/beacon/attestations.go:253]: this log will show-up at prysm runtime as 

```
[2020-05-11 20:27:06]  INFO blockchain: Processing block state transition...
[2020-05-11 20:27:06]  WARN: Got nil attestation
```
With no context of where it comes from or which function got a nil attestation. I would prefix rpc logs with the RPC prefix like we do in the other packages through a global: `var log = logrus.WithPrefix("rpc")`. Also, would add more informative error messages something like. log.Debug("Indexed attestations stream got a nil attestation") (we should make this one debug level).

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go]: thanks for the review

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go]: updated

**Pull request merged**